utils/device_pool.py

```diff
@@ -1,107 +1,211 @@
+        os.system("ANDROID_SERIAL=" + self.serial + " adb wait-for-device")
+        if search and search.group(0) >= '1.0.31':
+            ret = subprocess.call('/usr/bin/adb forward --list |grep ' + self.serial, shell=True, stdout=open(os.devnull, 'w'))
+            if ret == 0:
+                return True
+            else:
+                return False
+        else:
+            logger.error("adb forward --list not supported; recommend to upgrade 1.0.31 or newer version")
+            return False
+
+    def get_adb_forwarded_port(self):
+        out = exec_process(['/usr/bin/adb', 'version'])
+        search = re.search('[0-9\.]+', out)
+        os.system("ANDROID_SERIAL=" + self.serial + " adb wait-for-device")
+        if search and search.group(0) >= '1.0.31':
+            forward_list = exec_process(['/usr/bin/adb', 'forward', '--list']).splitlines()
+            for out in forward_list:
+                search_serial = re.search('\w+', out)
+                if search_serial and search_serial.group(0) == self.serial:
+                    logger.info("DeviceSerial[" + self.serial + "] forwarded")
+                    return int(re.search(' tcp:(\d+) ', out).group(1))
+        else:
+            logger.info("adb forward --list not supported; recommend to upgrade 1.0.31 or newer version")
         return None
 
-    def get_next_lock(self):
-        for next_index in range(self.current_lock_index + 1, len(self.lock_list)):
+    def acquire_file_lock(self):
+        self.lock_file.acquire(timeout=60)
+        return self.lock_file.path
+
+    def release_file_lock(self):
+        try:
+            self.lock_file.release()
+        except NotLocked:
+            logger.info("Lock is released")
+        self.lock_file = None
+
+    def create_adb_forward(self, specify_port=None):
+        if self.is_in_forwarded_list():
+            self.adb_forwarded_port = self.get_adb_forwarded_port()
+            logger.info("Using existing port [" + str(self.adb_forwarded_port) + "]")
+            return True
+        if specify_port:
+            self.adb_forwarded_port = specify_port
+        else:
+            self.adb_forwarded_port = self._find_available_port()
+        if self.serial and self.adb_forwarded_port:
+            ret = os.system("ANDROID_SERIAL=" + self.serial + " adb forward tcp:" + str(self.adb_forwarded_port) + " tcp:2828")
+            if ret != 0:
+                raise DMError("can't forward port to ANDROID_SERIAL[" + self.serial + "]")
+        logger.info("Port forwarding success serial:[" + self.serial + "] port:[" + str(self.adb_forwarded_port) + "]")
+        return True
+
+    def remove_adb_forward(self):
+        # Remove port forwarding
+        ret = os.system("ADNDROID_SERIAL=" + self.serial + " adb forward --remove tcp:" + str(self.adb_forwarded_port))
+        if ret != 0:
+            raise DMError("can't forward port to ANDROID_SERIAL[" + self.serial + "]")
+
+
+class DevicePool(object):
+
```

Comment:
Probably re-using upper-object makes it more elegant?
Then you can avoid writing
import device_pool.DevicePool
and turn to use
import device_pool